### DISCUSSION TITLE
Is there a way to apply arbitrary preprocessing to a 3D dataset?

### DISCUSSION THREAD

**Discussion by wayneadams** [aeon-toolkit/aeon#1492] [Q&A] [answered]
I see that the [TimeSeriesScaler](https://github.com/aeon-toolkit/aeon/blob/866c56f642c4fb55fdb667a43d003ed2b3a1be4a/aeon/transformations/collection/scaler.py#L12) applies sklearn's `StandardScaler` across each `(1_case, 1_channel, n_timepoints)` in a 3D dataset ("np-list", "numpy3D").

Is there an aeon-thonic way to apply arbitrary preprocessing functions to an arbitrarily subsets of channels? 

I'm envisioning something like sklearn's ColumnTransformer but for each timeseries in a 3D dataset.

I tried looking through the code, but I hit a cognitive wall with regards to multi-dimensional manipulation. Is this missing from aeon's API and should I open up a Feature Request + PR?

**Comment by baraline** [comment, marked as answer]
Hi @wayneadams, part of what you are mentioning is currently being worked on PR #1244. The goal of the PR is to create a broadcaster for series estimators (which works on single time series) to be able to use them as collections estimators  (which works on multiple time series) in your pipelines.

Concerning the selection of subset of channels, you would have to do it yourself, although a channel selection module is currently being worked on (see #1270).

While this is being worked on, I fear you would have to do some tinkering and looping yourself, happy to help if needed.

**Reply by wayneadams** [reply]
Solid. Thanks for the pointer to #1244. The discussion therein helped my understanding.

Though, I also fear that there's not a good human-understandable generic solution. 
>This should be the disclaimer for time series classification in general 😅

**Reply by baraline** [reply]
It's all also kinda related to the axis and input type problem : for example, is a 2D numpy array a collection of univariate time series `(n_cases, n_timestamps)` (we just expect 3D for collections now) or a multivariate time series `(n_channels, n_timestamps)` or `(n_timestamps, n_channels)` ? 

It caused a lot of pre/post-processing in the base classes that we are currently unifying in the library, so that we have clear guidelines on the expected format and on how users can deviate from the default and use their own. But we have a PR that is coming along for that, lots of things will become clearer thanks to that.

If at some point you didn't find a tool in aeon, don't hesitate to raise an issue, if you need it, others are likely to need it as well !
